File: src/context.ts
import { parse } from "std/path/parse.ts";
import { Asset, assetFromPath, buildCSS, PublicFile } from "./asset.ts";
import { PATHS, Paths } from "./constants.ts";
import { Content, contentFromPath } from "./content.ts";
import { walk } from "std/fs/walk.ts";
import { asyncToArray } from "./utils.ts";

export type Mode = "prod" | "dev";

export interface Metadata {
  url: URL;
  out: string;
  mode: Mode;
}

export interface Context {
  metadata: Metadata;
  mode: Mode;
  assets: Map<string, Asset>;
  pages: Map<string, Content>;
  public_files: Array<PublicFile>;
}

export const create_metadata = (mode: Mode): Metadata => {
  return {
    url: new URL(mode == "prod" ? "https://www.eons.io" : "http://localhost:3000"),
    out: PATHS.out,
    mode,
  };
};

async function* collectJs(paths: Paths): AsyncGenerator<Asset> {
  for await (const entry of Deno.readDir(paths.js)) {
    const path = `${paths.js}/${entry.name}`;
    yield assetFromPath(path);
  }
}

export const collectPages = async (paths: Paths): Promise<Map<string, Content>> => {
  const pages = new Array<Content>();

  for await (const entry of Deno.readDir(paths.pages)) {
    const path = `${paths.pages}/${entry.name}`;
    const content = await contentFromPath(path, "page");
    pages.push(content);
  }

  return new Map(pages.flat().map((page) => [parse(page.source).name, page]));
};

const collectPublicFiles = async (paths: Paths): Promise<Array<PublicFile>> => {
  const public_files = new Array<PublicFile>();
  for await (const entry of walk(paths.public, { includeDirs: false })) {
    public_files.push({ path: entry.path, prefix: paths.public });
  }

  return public_files;
};

export const createContext = async (paths: Paths, mode: Mode): Promise<Context> => {
  const metadata = create_metadata(mode);
  const assets = new Map<string, Asset>();
  const public_files = await collectPublicFiles(paths);

  assets.set("styles.css", await buildCSS(paths, mode));
  (await asyncToArray(collectJs(paths))).map((a) => assets.set(a.filename, a));

  const pages = await collectPages(paths);

  return {
    metadata,
    mode,
    assets,
    pages,
    public_files,
  };
};

File: src/context.rs
use crate::{
    asset::{Asset, PublicFile},
    constants::*,
    content::Content,
    Mode,
};
use ahash::AHashMap;
use anyhow::Result;
use minijinja::{path_loader, Environment};
use minijinja_autoreload::AutoReloader;
use std::path::PathBuf;
use url::Url;

#[derive(Debug)]
pub struct Metadata {
    pub url: Url,
    pub out: PathBuf,
}

impl Metadata {
    pub fn new(mode: Mode) -> Result<Self> {
        Ok(Self {
            url: match mode {
                Mode::Prod => Url::parse("https://www.eons.io")?,
                Mode::Dev => Url::parse("http://localhost:3000")?,
            },
            out: PathBuf::from(OUT_PATH),
        })
    }
}

pub struct Context {
    pub metadata: Metadata,
    pub assets: AHashMap<String, Asset>,
    pub pages: AHashMap<String, Content>,
    pub templates: AutoReloader,
    pub public_files: Vec<PublicFile>,
    pub mode: Mode,
}

impl Context {
    pub fn new(
        paths: &Paths,
        metadata: Metadata,
        assets: AHashMap<String, Asset>,
        pages: AHashMap<String, Content>,
        public_files: Vec<PublicFile>,
        mode: Mode,
    ) -> Self {
        let template_path = paths.templates.clone();
        let env = AutoReloader::new(move |notifier| {
            let mut env = Environment::new();
            env.set_loader(path_loader(&template_path));

            notifier.set_fast_reload(true);

            notifier.watch_path(&template_path, true);
            Ok(env)
        });

        Self {
            metadata,
            assets,
            pages,
            templates: env,
            public_files,
            mode,
        }
    }

    pub fn _update_asset(&mut self, path: impl Into<String>, asset: Asset) {
        self.assets.insert(path.into(), asset);
    }

    pub fn _update_page(&mut self, path: impl Into<String>, page: Content) {
        self.pages.insert(path.into(), page);
    }
}
